### DISCUSSION TITLE
NUOPC interface for coupling

### DISCUSSION THREAD

**Discussion by uturuncoglu** [NCAR/noahmp#119] [General]
@cenlinhe I am trying to port NUOPC coupling interface to new restructured code. I could not be sure about how I need to set TemperatureSoilBottom variable that is in the forcing data structure. I just wonder if you could give more information about it. How are you handling it in ERA5 case?

**Comment by cenlinhe** [comment]
The TemperatureSoilBottom variable is the deep soil temperature as a bottom boundary condition to force soil temperature calculations. This variable is used only if the noahmp physics option TBOT_OPTION = 2. This variable is typically set to the annual mean surface/ground temperature.

**Reply by uturuncoglu** [reply]
@cenlinhe Thanks. I fixed it.

**Comment by uturuncoglu** [comment]
@barlage @cenlinhe I have couple of questions,

- Is soil color used by the restructured version of NoahMP. I checked but I could not find. It seems that it is used by `transfer_mp_parameters()` call as a parameter in the old code. 
- `snoalb` seems not used by the old model but it is still an argument to CCPP noahmp driver. Why we have this variable even if it is not used. Maybe I am missing something in here.
- I found the that `shdmax` in the old code is `VegFracAnnMax` in new one but I could not find `shdmin`. Is it still used in the restructured code.

**Reply by cenlinhe** [reply]
1. The 2D soil color map is currently not used in refactored NoahMP (no I/O code capability developed yet, will do it later this year). By default, the refactored NoahMP uses a 8-class soil color specified in NoahmpTable.TBL, and the soil color is set to a fixed value (=4) everywhere: https://github.com/NCAR/noahmp/blob/master/drivers/hrldas/ConfigVarInTransferMod.F90#L66
2. snoalb is not used by NoahMP and I believe I removed it from the refactored NoahMP. The CCPP version is a legacy of the old code without clean-up.
3. shdmin is not used by NoahMP. It is not used in refactored NoahMP.

**Reply by uturuncoglu** [reply]
@cenlinhe Thanks for the clarification. Another question is that which variables need to be read for the initial condition except forcing and other fixed information like slope, vegetation type.

**Reply by uturuncoglu** [reply]
@cenlinhe In the old version of code, I was reading followings,

```
sheleg - snow water equivalent
snwdph - snow depth
canopy - canopy water
tsea - skin temperature
stc - soil temperature at layers
smc - soil moisture at layers
slc - soil liquid at layers
zorl - surface roughness
uustar - friction velocity
vtype - vegetation type
```

It is also important to know which variable need to be read for the exact restart. I was reading in all the variables available to noahmp in the past for that case.

**Reply by cenlinhe** [reply]
you can look at this HRLDAS I/O module for the variables (some I/O modules and variables are only for agriculture and crop, which you do not need to include because UFS will not activate those Noah-MP physics options at this stage):
https://github.com/NCAR/hrldas/blob/master/hrldas/IO_code/module_hrldas_netcdf_io.F#L1697
https://github.com/NCAR/hrldas/blob/master/hrldas/IO_code/module_hrldas_netcdf_io.F#L238
https://github.com/NCAR/hrldas/blob/master/hrldas/IO_code/module_hrldas_netcdf_io.F#L405
https://github.com/NCAR/hrldas/blob/master/hrldas/IO_code/module_hrldas_netcdf_io.F#L512
https://github.com/NCAR/hrldas/blob/master/hrldas/IO_code/module_hrldas_netcdf_io.F#L575
https://github.com/NCAR/hrldas/blob/master/hrldas/IO_code/module_hrldas_netcdf_io.F#L665
https://github.com/NCAR/hrldas/blob/master/hrldas/IO_code/module_hrldas_netcdf_io.F#L756

**Reply by uturuncoglu** [reply]
@cenlinhe Thanks. That is really helpful. Let me check those. Sometime it is hard to find the relevant information in those codes.

**Reply by cenlinhe** [reply]
For restart variables, look at this:
https://github.com/NCAR/hrldas/blob/master/hrldas/IO_code/module_NoahMP_hrldas_driver.F#L331-L420

**Comment by uturuncoglu** [comment]
@barlage We might need to create new initial condition for C96 configuration coupled with ERA5 since I think the current version used with old version of NoahMP has no all the variables required. Here is the content of the current file. Anyway, let me know what you think. At this point I am implementing a routine that will get exiting fields but maybe we need more.

```
	float time(time) ;
		time:long_name = "time" ;
		time:units = "seconds since 1970-01-01 00:00:00" ;
		time:_FillValue = 9.96921e+36f ;
	float geolat(lat, lon) ;
		geolat:long_name = "latitude" ;
		geolat:units = "degrees_north" ;
		geolat:_FillValue = 9.96921e+36f ;
	float geolon(lat, lon) ;
		geolon:long_name = "longitude" ;
		geolon:units = "degrees_east" ;
		geolon:_FillValue = 9.96921e+36f ;
	float snow_water_equivalent(time, lat, lon) ;
		snow_water_equivalent:long_name = "snow water equivalent" ;
		snow_water_equivalent:units = "mm" ;
		snow_water_equivalent:_FillValue = 9.96921e+36f ;
	float snow_depth(time, lat, lon) ;
		snow_depth:long_name = "snow depth" ;
		snow_depth:units = "m" ;
		snow_depth:_FillValue = 9.96921e+36f ;
	float canopy_water(time, lat, lon) ;
		canopy_water:long_name = "canopy surface water" ;
		canopy_water:units = "mm" ;
		canopy_water:_FillValue = 9.96921e+36f ;
	float skin_temperature(time, lat, lon) ;
		skin_temperature:long_name = "surface skin temperature" ;
		skin_temperature:units = "K" ;
		skin_temperature:_FillValue = 9.96921e+36f ;
	float soil_temperature(time, soil_levels, lat, lon) ;
		soil_temperature:long_name = "soil temperature" ;
		soil_temperature:units = "K" ;
		soil_temperature:_FillValue = 9.96921e+36f ;
	float soil_moisture(time, soil_levels, lat, lon) ;
		soil_moisture:long_name = "volumetric soil moisture" ;
		soil_moisture:units = "m3/m3" ;
		soil_moisture:_FillValue = 9.96921e+36f ;
	float soil_liquid(time, soil_levels, lat, lon) ;
		soil_liquid:long_name = "volumetric soil liquid" ;
		soil_liquid:units = "m3/m3" ;
		soil_liquid:_FillValue = 9.96921e+36f ;
```

**Reply by uturuncoglu** [reply]
@cenlinhe What is the equivalent of WEASD?

**Reply by cenlinhe** [reply]
this is the snow water equivalent (SWE), unit: mm

**Reply by uturuncoglu** [reply]
@cenlinhe I could not find it in the excel file. Maybe it was missing.

**Reply by cenlinhe** [reply]
sorry for the confusion. this is a temporary HRLDAS variable in the I/O subroutine declaration, when calling this subroutine, the variable passed to this subroutine is actually NoahmpIO%SNOW.

**Reply by uturuncoglu** [reply]
@cenlinhe Sorry but i could not find any variable called SWE in the model side (https://github.com/search?q=repo%3ANCAR%2Fnoahmp%20SWE&type=code). Which variable in the model needs to be assigned for it?

**Reply by cenlinhe** [reply]
Here it is: https://github.com/NCAR/noahmp/blob/master/drivers/hrldas/WaterVarInTransferMod.F90#L54

**Reply by uturuncoglu** [reply]
@cenlinhe Okay. Thanks. It is very tricky to find those variables if you are switching from old version of code to new version. It is something like rabbit hole.

**Comment by uturuncoglu** [comment]
@cenlinhe @barlage In the old version of the code STC dimension was equal to number of soil layer (https://github.com/NOAA-EMC/noahmp/blob/0cd3e23ae5d35ef93e205e4d0c3b13ccc0d851f5/drivers/ccpp/noahmpdrv.F90#L778) but in the new code TemperatureSoilSnow also includes the snow layers. How we need to set this new variable? @barlage The initial condition has only soil levels and does not include snow layers.

**Reply by cenlinhe** [reply]
In the old NoahMP version, STC still has a dimension of soil+snow layers:
https://github.com/NCAR/noahmp/blob/noahmp_v4.5_develop/drivers/wrf/module_sf_noahmpdrv.F#L804-L805

**Reply by uturuncoglu** [reply]
@cenlinhe Thanks for your help. Okay. It is more clear to me since it is utilized with two different variable. I need to check CCPP driver to see how it is handled over there. Thanks again and sorry for lots of questions.

**Comment by barlage** [comment]
@uturuncoglu we should talk about some of the more general plans here. All of these initial states are created in [ccpp](https://github.com/ufs-community/ccpp-physics/blob/9b0ac7b16a45afe5e7f1abf9571d3484158a5b43/physics/Interstitials/UFS_SCM_NEPTUNE/GFS_phys_time_vary.fv3.F90#L406). So, we should try to come up with a generalization of this. There shouldn't be much that has changed in regards to initialized variables between the refactored and non-refactored code. I had wanted the code that is referenced above to be in the noahmp part of ccpp.

**Reply by uturuncoglu** [reply]
@barlage Yes, that would be great. I am not sure at this point I could make new version of code available to UFS. There are lots of tiny details. The core infrastructure is ready but I need to fill all the model variables in a correct way. So, referencing old model, new version and hrldas driver for each variable is taking too much time.

**Comment by uturuncoglu** [comment]
@cenlinhe There is a variable in noahmp table `noahmp_crop_parameters` section as `MPI`. This has a special meaning and used as CPP flag (again `MPI`). So, when you compile code under UFS this is replaced with 1 and then it creates error. I might need to change the name of the variable to something else in the table file and also in the routine that defines the namelist in the NUOPC cap. Do you have any idea?

**Reply by cenlinhe** [reply]
what specific variable? do you have the line of the code as an example in the current noahmp? I am not very familiar with the MPI and CPP treatment.

**Reply by uturuncoglu** [reply]
It is MPI variable in this line. https://github.com/NCAR/noahmp/blob/44bef284e62004f911749ee41daef5778f042bf4/drivers/hrldas/NoahmpReadTableMod.F90#L134
There is a CPP flags like -DMPI in the build and since that matches with the MPI variable name. It replaces it with 1 before compiling and create issue in my side.

**Reply by cenlinhe** [reply]
Hmm.... I see. I didn't expect that when naming those variables. maybe the best way is to change the table variable name for "MPI" to "MPII"? If so, I will need to change it in the HRLDAS and Noah-MP side too in the GitHub to keep consistent code. Let me know what you think? Are you currently linking UFS/Noah-MP part directly to our community GitHub through submodule?

**Reply by uturuncoglu** [reply]
@cenlinhe yes, I agree. Let me fix in my side. If I could finalize the coupling interface and create PR in NoahMP repository we could make changes in the HRLDAS too. I am also plaining to create issue in NoahMP repository side to keep track it. Thanks for your help.

**Reply by cenlinhe** [reply]
sounds good. thanks!
